Github CI tests

elizabethmcd opened the pull request:
This PR adds support for Github CI for linting checks and tests of both the Illumina and Nanopore workflows. Review #31 before this PR.

elizabethmcd commented:
I can't confirm if these will work I think until they are merged and then run the tests on future commits?

taylorreiter commented on .github/ISSUE_TEMPLATE/bug_report.yml:
```diff
@@ -0,0 +1,47 @@
+    attributes:
+      label: System information
+      description: |
+        * Nextflow version _(eg. 21.10.3)_
```
```suggestion
        * Nextflow version _(eg. 22.10.1)_
```

elizabethmcd replied:
ah poop I forgot you had to make these corresponding changes

taylorreiter commented on .github/awstest.yml:
```diff
@@ -0,0 +1,29 @@
+            {
+              "outdir": "s3://${{ secrets.AWS_S3_BUCKET }}/metagenomics/results-${{ github.sha }}"
+            }
+          profiles: test_illumina,test_nanopore,aws_tower
```
Oooh can profiles be combined like this? would this submit two different runs to tower, one for illumina and one for nanopore, or is it just one? it would be really cool if `test_illumina` output a different multiqc than `test_nanopore`, but I'm guessing it doesnt.

taylorreiter replied:
@mertcelebi do you think we could do this for the tower full data sets over at seqqc?

mertcelebi replied:
Great question! We should test, I am not sure TBH. ie I am not sure how the combined profiles work.

elizabethmcd replied:
I'm not sure either, other than I know I would like to test the separate profiles. And yes they do output different multiqc reports, I've already tested that configuration!

elizabethmcd replied:
for now I've changed so it just has `test_illumina` although I would like to test both in the future, but starting simple for now

taylorreiter commented on .github/awstest.yml:
```diff
@@ -0,0 +1,29 @@
+          workdir: s3://${{ secrets.AWS_S3_BUCKET }}/work/metagenomics/work-${{ github.sha }}
+          parameters: |
+            {
+              "outdir": "s3://${{ secrets.AWS_S3_BUCKET }}/metagenomics/results-${{ github.sha }}"
```
do you want the outdir to have something like `test_` in the name as well? it might help if we start running this workflow a bunch on tower with real data

taylorreiter commented on .github/ci.yml:
```diff
@@ -0,0 +1,61 @@
+        run: |
+          nextflow run ${GITHUB_WORKSPACE} -profile test_illumina,docker --outdir ./results_illumina
+
+  test_nanopore:
+    name: Run Nanopore workflow with test data
+    # Only run on push if this is the nf-core dev branch (merged PRs)
+    if: "${{ github.event_name != 'push' || (github.event_name == 'push' && github.repository == 'Arcadia-Science/metagenomics') }}"
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        NXF_VER:
+          - "22.10.1"
+          - "latest-everything"
+    steps:
+      - name: Check out pipeline code
+        uses: actions/checkout@v3
+
+      - name: Install Nextflow
+        uses: nf-core/setup-nextflow@v1
+        with:
+          version: "${{ matrix.NXF_VER }}"
+
+      - name: Run Nanopore workflow with test data
+        run: |
+          nextflow run ${GITHUB_WORKSPACE} -profile test_nanopore,docker --outdir ./results_nanopore
```
can you use a matrix or something here so that you only have to do all of the test setup once?

it looks like the taxprofiler repo just copies like you have: https://github.com/nf-core/taxprofiler/blob/dev/.github/workflows/ci.yml. As does rnaseq: https://github.com/nf-core/rnaseq/blob/master/.github/workflows/ci.yml. So maybe this is fine the way it is!

taylorreiter commented on .github/linting.yml:
```diff
@@ -0,0 +1,61 @@
+  EditorConfig:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v3
```

taylorreiter commented on .github/linting.yml:
```diff
@@ -0,0 +1,61 @@
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: actions/setup-node@v2
```
```suggestion
      - uses: actions/setup-node@v3
```

taylorreiter commented on .github/linting.yml:
```diff
@@ -0,0 +1,61 @@
+  Prettier:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v3
```

taylorreiter commented on .github/linting.yml:
```diff
@@ -0,0 +1,61 @@
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: actions/setup-node@v2
```
```suggestion
      - uses: actions/setup-node@v3
```

taylorreiter commented on .github/linting.yml:
```diff
@@ -0,0 +1,61 @@
+  PythonBlack:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
```suggestion
      - uses: actions/checkout@v3
```

taylorreiter commented on .github/linting.yml:
```diff
@@ -0,0 +1,61 @@
+            We highly recommend setting up Black in your code editor so that this formatting is done automatically on save. Ask about it on Slack for help!
+            Thanks again for your contribution!
+          repo-token: ${{ secrets.GITHUB_TOKEN }}
+          allow-repeats: false
```
do you want to do nf-core linting? something like this:
```
  nf-core:
    runs-on: ubuntu-latest
    steps:
      - name: Check out pipeline code
        uses: actions/checkout@v3

      - name: Install Nextflow
        uses: nf-core/setup-nextflow@v1

      - uses: actions/setup-python@v4
        with:
          python-version: "3.7"
          architecture: "x64"

      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install nf-core

      - name: Run nf-core lint
        env:
          GITHUB_COMMENTS_URL: ${{ github.event.pull_request.comments_url }}
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          GITHUB_PR_COMMIT: ${{ github.event.pull_request.head.sha }}
        run: nf-core -l lint_log.txt lint --dir ${GITHUB_WORKSPACE} --markdown lint_results.md

      - name: Save PR number
        if: ${{ always() }}
        run: echo ${{ github.event.pull_request.number }} > PR_number.txt

      - name: Upload linting log file artifact
        if: ${{ always() }}
        uses: actions/upload-artifact@v3
        with:
          name: linting-logs
          path: |
            lint_log.txt
            lint_results.md
            PR_number.txt
```

elizabethmcd replied:
we have this commented out in the cookiecutter repo at the moment and I'm not sure if it's because it doesn't work or for some other reason - does it work for you?

taylorreiter replied:
it works on seqqc, but im not sure if it would work with metagenomics. if you have nf-core tools installed locally, I think you could run `nf-core lint` and see if it your pipeline passes. I have a bunch of ignores set up for specific files that I expect to deviate from the linter, so it did take some config, but its also caught some stuff for me in the past that has been useful

elizabethmcd replied:
I'm going to leave this out for now but I will make an issue to remind me to come back to this

taylorreiter reviewed (commented):
I left some comments, suggestions, and questions. Commenting for now because you need to bump the nextflow version in the README and the nextflow.config (line 180) to match the nextflow version you're running on CI.

mertcelebi commented on subworkflows/local/nanopore_mapping_depth.nf:
```diff
@@ -0,0 +1,41 @@
+// Minimap2 subworkflow to index an assembly and map reads to the assembly
```
Why are these changes here?

mertcelebi commented:
@elizabethmcd can you fix this PR so it only has the GitHub CI changes and not the changes from your other PR?

elizabethmcd commented:
I have fixed the PR so it doesn't have the files from the previous PR to re-review. I've also changed the `awstest.yml` test to have just the `test_illumina` profile. I've looked at other nf-core workflows that are similarly setup with two separate "subworkflows" and they have a general `test` profile that really only tests one of the subworkflows for this purpose. In the future it would be nice to test both, but for the purposes of getting this up and going I think this is fine.

taylorreiter reviewed (approved):
LGTM! I checked it against what I do in seqqc so I think it should work, but always a chance I missed something. I'll be excited to see if it works!
